### PR TITLE
Use JQuery to solve Catastrophic Backtracking with regex

### LINKED ISSUE #39101: Challenge freezing on incorrect input
**Describe your problem and how to reproduce it:**
The Define the Head and Body of an HTML Document challenge freezes if you enter in code like

```html
<!DOCTYPE html>
<html>
  <head>
    <title>The best page ever</title>
  <head>

  <body>
    <h1>The best page ever</h1>
    <p>Cat ipsum dolor sit amet, jump </p>
  <body> <!-- this tag should be closing -->
</html>
```

where the `body` tag is not closed.  This is a common enough mistake and the freezing is not directly caused by the user code - it's how our tests interact with it.

**Add a Link to the page with the problem:**

https://www.freecodecamp.org/learn/responsive-web-design/basic-html-and-html5/define-the-head-and-body-of-an-html-document

**The cause**

The last test causes the problem, I think the regex gets stuck in a loop.

### PR DESCRIPTION
**Pull request by PSN221B**
Checklist:

- [x ] I have read [freeCodeCamp's contribution guidelines](https://contribute.freecodecamp.org).
- [x ] My pull request has a descriptive title (not a vague title like `Update index.md`)
- [x ] My pull request targets the `master` branch of freeCodeCamp.
- [x ] All the files I changed are in the same world language, for example: only English changes, or only Chinese changes, etc.

Closes #39101 

I have used the same logic as the first 3 test cases to check the code using JQuery.

### CODE REVIEW

File: curriculum/challenges/english/01-responsive-web-design/basic-html-and-html5/define-the-head-and-body-of-an-html-document.english.md [hunk 1]
```diff
@@ -46,10 +46,9 @@ tests:
   - text: The <code>body</code> element should be a child of the <code>html</code> element.
     testString: assert($('html').children('body').length == 1);
   - text: The <code>head</code> element should wrap around the <code>title</code> element.
-    testString: assert(code.match(/<head>\s*?<title>\s*?.*?\s*?<\/title>\s*?<\/head>/gi));
+    testString: assert($('head').children('title').length == 1);
```
Comment (thecodingaviator) [inline curriculum/challenges/english/01-responsive-web-design/basic-html-and-html5/define-the-head-and-body-of-an-html-document.english.md:49]: This test is failing

**Review by thecodingaviator — DISMISSED** [review]

File: curriculum/challenges/english/01-responsive-web-design/basic-html-and-html5/define-the-head-and-body-of-an-html-document.english.md (hunk 1 above)
Comment (PSN221B) [reply curriculum/challenges/english/01-responsive-web-design/basic-html-and-html5/define-the-head-and-body-of-an-html-document.english.md:49]: But the logic is correct. I don't know how the test files are being converted. I guess it has something to do with that. I am using the same JQuery concept that the upper test cases have used. 
See issue #39111

**Comment by lasjorg** [comment]
The title element is ending up inside the body element for some reason. Not really sure what is going on.
title-element

**Comment by lasjorg** [comment]
Am I wrong or was the title never in the head and we just didn't see it because the regex is testing the code and not the actual DOM output.

https://github.com/freeCodeCamp/freeCodeCamp/blob/master/client/src/templates/Challenges/rechallenge/builders.js

**Comment by ojeytonwilliams** [comment]
@lasjorg you're right, to my knowledge the head and body are always silently removed.

**Comment by thecodingaviator** [comment]
> The title element is ending up inside the body element for some reason. Not really sure what is going on.
> title-element

Good find.

File: curriculum/challenges/english/01-responsive-web-design/basic-html-and-html5/define-the-head-and-body-of-an-html-document.english.md (hunk 1 above)
Comment (thecodingaviator) [reply curriculum/challenges/english/01-responsive-web-design/basic-html-and-html5/define-the-head-and-body-of-an-html-document.english.md:49]: Yup. See below

**Comment by RandellDawson** [comment]
@Manish-Giri @thecodingaviator @Sky020 @nibble0101 Can we get two of you to review this PR since @moT01 and I added 3 extra commits to fixed the current bug using regex in the tests.  Once @ojeytonwilliams fixes the overall bug causes the original tests to fail, we can revisit using jQuery again.

**Review by Manish-Giri — APPROVED** [review]
LGTM. Tested out various combinations, and the regex seems to catch all of them.

One thing I noticed, if the `<head>...</head>` tags are present **inside** the `<body>...</body>` tags (as shown below), the tests still pass. Shouldn't be a blocker but just wanted to point it out -
```html
<html>
  <body>
  <head>
    <title>The best page ever</title>
  </head>
  <h1>The best page ever</h1>
  <p>...</p>
</body>
</html>
```

**Comment by RandellDawson** [comment]
> LGTM. Tested out various combinations, and the regex seems to catch all of them.
> 
> One thing I noticed, if the `<head>...</head>` tags are present **inside** the `<body>...</body>` tags (as shown below), the tests still pass. Shouldn't be a blocker but just wanted to point it out -
> 
> ```
> <html>
>   <body>
>   <head>
>     <title>The best page ever</title>
>   </head>
>   <h1>The best page ever</h1>
>   <p>...</p>
> </body>
> </html>
> ```

Yeah there are a few edge cases that might be difficult to detect,  I think even with the jQuery selector tests, it would not have caught something like this.  We would have to explicitly write such a test the would need to fail if they did it.  I have some tests like that on the new project-based curriculum we are working on.

**Review by ShaunSHamilton — APPROVED** [review]
LGTM 👍 This resolves the linked issue. So, I am happy with it.

As @Manish-Giri mentioned, there are a few edge cases which should not pass, but do.

**Comment by RandellDawson** [comment]
Would anyone like to create a GitHub issue for the edge case mentioned above?  We could add a `help wanted` and have other contributors work on it.

**Comment by nhcarrigan** [comment]
@RandellDawson I can create an issue for this, sure. 🙂

**Comment by nhcarrigan** [comment]
image

Actually, that specific edge case mentioned appears to error out the last test.

**Comment by RandellDawson** [comment]
@nhcarrigan Add that to the issue too.
